### PR TITLE
Fixing fast connections bug AASTMIP-4199

### PR DESCRIPTION
**Pull request by golemic**
Fixing fast connections bug AASTMIP-4199

### CODE REVIEW

File: src/coap_main.c [hunk 1]
```diff
@@ -365,30 +365,38 @@ CoAP_Result_t _rom CoAP_handleDisconnectEvt(uint32_t transport_ctx) {
...
+		if (res != COAP_OK) {
+			return res;
+		}
 
-	/* Delete/abort any pending notification interaction. */
-	res = CoAP_RemoveObserverInteractions(pObserver);
-	if (res != COAP_OK) {
-		return res;
+		/* Remove observer from the resource. */
+		CoAP_RemoveObserverFromResource(&pRes->pListObservers,
```
Comment (magdalenaszumny) [inline src/coap_main.c]: Should the return value be checked?

File: src/coap_main.c [hunk 2]
```diff
@@ -365,30 +365,38 @@ CoAP_Result_t _rom CoAP_handleDisconnectEvt(uint32_t transport_ctx) {
...
-										  &pObserver->Ep,
-										  pObserver->Token);
-	if (res != COAP_REMOVED) {
-		return res;
+	
+	CoAP_Interaction_t *pIA = CoAP.pInteractions;
+	while ( NULL != pIA)
+	{
+		CoAP_Interaction_t *nextPIA = pIA->next;
+		if ( pIA->RemoteEp.transport_ctx == transport_ctx)
```
Comment (magdalenaszumny) [inline src/coap_main.c]: I think that formatting should be (to be consistent with code in this repo):
```suggestion
		if (pIA->RemoteEp.transport_ctx == transport_ctx)
```

File: src/coap_main.c [hunk 3]
```diff
@@ -365,30 +365,38 @@ CoAP_Result_t _rom CoAP_handleDisconnectEvt(uint32_t transport_ctx) {
...
-	/* Remove observer from the resource. */
-	res = CoAP_RemoveObserverFromResource(&pRes->pListObservers,
-										  pObserver->socketHandle,
-										  &pObserver->Ep,
-										  pObserver->Token);
-	if (res != COAP_REMOVED) {
-		return res;
+	
+	CoAP_Interaction_t *pIA = CoAP.pInteractions;
+	while ( NULL != pIA)
```
Comment (magdalenaszumny) [inline src/coap_main.c]: I think that formatting should be (to be consistent with code in this repo):
```suggestion
	while (NULL != pIA)
```

File: src/coap_main.c [hunk 4]
```diff
@@ -365,30 +365,38 @@ CoAP_Result_t _rom CoAP_handleDisconnectEvt(uint32_t transport_ctx) {
...
-	/* Delete/abort any pending notification interaction. */
-	res = CoAP_RemoveObserverInteractions(pObserver);
-	if (res != COAP_OK) {
-		return res;
+		/* Remove observer from the resource. */
+		CoAP_RemoveObserverFromResource(&pRes->pListObservers,
+											pObserver->socketHandle,
+											&pObserver->Ep,
+											pObserver->Token);
 	}
```
Comment (magdalenaszumny) [inline src/coap_main.c:386]: I don't have enough understanding of this issue, so just to confirm - if CoAP_FindObserverAndResourceByTransportCtx returns something else than COAP_OK it is ok to continue execution of this function?

**Review by magdalenaszumny — COMMENTED** [review]
Nice!

**Comment by magdalenaszumny** [comment]
It would be perfect, if you could add some more explanation, what exactly was the issue and how it was fixed :)

File: src/coap_main.c (hunk 4 above)
Comment (golemic) [reply src/coap_main.c:386]: This is the main part of this issue workaround. In fast connections tests we are unable to delete some old interaction due to lack of information about observers in the list - so it was a situation with observation without observer so those function normally are ok but in this edge case we need to manually erase interactions.

**Review by magdalenaszumny — APPROVED** [review]

**Pull request merged**